File: Sources/BLE/ProgressorClient.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import Foundation

/// What the app knows about the link to the gauge.
enum ProgressorConnectionState: Sendable, Equatable {
    /// Nothing attempted yet — no CBCentralManager exists, so no permission prompt
    /// has been shown.
    case idle
    case bluetoothOff
    case unauthorized
    /// No BLE hardware. This is what the Simulator always reports, which is why the
    /// mock client exists.
    case unsupported
    case scanning
    case connecting
    case connected
    case disconnected(reason: String?)

    var isConnected: Bool { self == .connected }

    var isBusy: Bool {
        switch self {
        case .scanning, .connecting: true
        default: false
        }
    }

    /// User-facing one-liner for the device chip.
    var label: String {
        switch self {
        case .idle: String(localized: "Not connected")
        case .bluetoothOff: String(localized: "Bluetooth off")
        case .unauthorized: String(localized: "Bluetooth denied")
        case .unsupported: String(localized: "No Bluetooth")
        case .scanning: String(localized: "Searching…")
        case .connecting: String(localized: "Connecting…")
        case .connected: String(localized: "Connected")
        case .disconnected(let reason): reason ?? String(localized: "Disconnected")
        }
    }
}

/// Why a stream start was requested. The cause stays attached if the write has to wait
/// behind a tare, because a later write without its original reason makes the next
/// hardware trace impossible to interpret.
enum StreamStartCause: String, Sendable, CaseIterable {
    case initial
    case reconnect
    case foreground
    case watchdog
    case tareRecovery
    case manualMeasurement
    /// The human tapped Wake. Distinct from `.watchdog` on purpose: attributing a manual
    /// recovery to the automatic one would make the breadcrumb report lie about who
    /// revived the stream, in the one log that exists to explain exactly that.
    case manualWake

    var label: String {
        switch self {
        case .initial: "initial"
        case .reconnect: "reconnect"
        case .foreground: "foreground"
        case .watchdog: "watchdog"
        case .tareRecovery: "tare recovery"
        case .manualMeasurement: "manual measurement"
        case .manualWake: "manual wake"
        }
    }
}

/// Why a stream stopped. The counterpart to `StreamStartCause`, and the reason the
/// breadcrumb ring can tell a genuine stall from the app doing exactly what it should.
/// **Deliberately has no default anywhere it is taken.** A defaulted cause is a cause
/// that is wrong at the call sites nobody revisited, and a diagnostic that quietly
/// mislabels an automatic stop as "stopped by hand" is worse than one that says nothing:
/// it sends whoever reads the log looking for a user who was never there.
enum StreamStopCause: String, Sendable, CaseIterable {
    case sessionEnded
    case userStopped
    case measurementComplete
    case screenClosed
    case timedOut
    case disconnecting
    case sleeping

    var label: String {
        switch self {
        case .sessionEnded: "session ended"
        case .userStopped: "stopped by hand"
        case .measurementComplete: "measurement finished"
        case .screenClosed: "screen closed"
        case .timedOut: "timed out"
        case .disconnecting: "disconnecting"
        case .sleeping: "gauge sleeping"
        }
    }
}

/// Client-side lifecycle facts that are not themselves connection-state changes.
/// They let the store's in-memory breadcrumb ring include quarantine boundaries without
/// passing CoreBluetooth objects out of the client.
enum ProgressorClientDiagnostic: Sendable {
    case retiringPeripheral
    case quarantineReleased
    case streamStartDeferred(StreamStartCause)
    case streamStartWritten(StreamStartCause)
}

/// The seam between the app and the gauge.
///
/// Marked `@MainActor` on the PROTOCOL so isolation propagates to every conformer:
/// the live client passes `queue: .main` to CoreBluetooth, so its delegate callbacks
/// genuinely arrive on the main thread and there is no isolation boundary to cross
/// anywhere in the stack. At ~10 notifications/sec of microsecond-cheap TLV decoding
/// there is nothing to gain from a second isolation domain, and plenty to lose.
///
/// Callback-style rather than `AsyncStream` for the same reason — the store wires
/// `onEvent` the way the sibling apps wire their session callbacks, and every hop
/// stays visible in one place.
@MainActor
protocol ProgressorClient: AnyObject {
    var onEvent: ((ProgressorEvent) -> Void)? { get set }
    var onPacketBoundary: ((PacketBoundary) -> Void)? { get set }
    var onStateChange: ((ProgressorConnectionState) -> Void)? { get set }
    var onDiagnostic: ((ProgressorClientDiagnostic) -> Void)? { get set }

    var state: ProgressorConnectionState { get }
    /// Advertised name of the connected unit, e.g. `Progressor_1234`.
    var deviceName: String? { get }
    /// Which physical device family this client drives. Capability gating — tare
    /// behaviour, background streaming, timing source — reads `kind.capabilities`,
    /// never the concrete client type.
    var kind: GaugeKind { get }

    /// Scan for, and connect to, the first Progressor found. Idempotent.
    func connect()
    func disconnect()
    func send(_ command: ProgressorCommand)
    func startStreaming(cause: StreamStartCause)
    /// Ask the gauge to sleep and end this connection without allowing an automatic
    /// reconnect to race the power-down command.
    func sleepDevice()
}

enum PacketBoundary {
    case began(receivedAt: TimeInterval)
    case ended
}

extension ProgressorClient {
    // Existing test/third-party adapters can continue to emit individual events.
    var onPacketBoundary: ((PacketBoundary) -> Void)? {
        get { nil }
        set {}
    }

    /// The default covers the two existing clients (live Tindeq and the mock,
    /// which scripts a Tindeq); multi-device clients override it.
    var kind: GaugeKind { .progressor }

    func tare() { send(.tare) }
    func stopStreaming() { send(.stopWeightMeasurement) }
    func readBattery() { send(.getBatteryVoltage) }
    /// Explicit power-down only; normal session completion keeps the gauge available
    /// for the second daily session.
    func sleepDevice() { send(.enterSleep) }
}

File: android/app/src/main/kotlin/run/nuri/getagrip/ble/ProgressorClient.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ble

import run.nuri.getagrip.engine.GaugeKind
import run.nuri.getagrip.engine.L10n
import run.nuri.getagrip.engine.ProgressorCommand
import run.nuri.getagrip.engine.ProgressorEvent

/// What the app knows about the link to the gauge.
///
/// TRANSLATION NOTE: Swift's `enum` becomes a sealed interface (as `ProgressorEvent` in
/// `:engine`). `data object`/`data class` give structural equality; thread confinement is a
/// contract stated on `ProgressorClient`, not compiler-checked.
sealed interface ProgressorConnectionState {
    /// Nothing attempted yet — no scanner or GATT client, so no permission prompt shown.
    data object Idle : ProgressorConnectionState

    data object BluetoothOff : ProgressorConnectionState

    data object Unauthorized : ProgressorConnectionState

    /// No BLE hardware. TRANSLATION NOTE: what the iOS Simulator always reports; an Android
    /// emulator has no working BLE either, so `MockProgressorClient` is reached the same
    /// two ways.
    data object Unsupported : ProgressorConnectionState

    data object Scanning : ProgressorConnectionState

    data object Connecting : ProgressorConnectionState

    data object Connected : ProgressorConnectionState

    data class Disconnected(val reason: String? = null) : ProgressorConnectionState

    val isConnected: Boolean get() = this is Connected

    val isBusy: Boolean get() = this is Scanning || this is Connecting

    /// User-facing one-liner for the device chip.
    val label: String
        get() = when (this) {
            Idle -> L10n.tr("Not connected")
            BluetoothOff -> L10n.tr("Bluetooth off")
            Unauthorized -> L10n.tr("Bluetooth denied")
            Unsupported -> L10n.tr("No Bluetooth")
            Scanning -> L10n.tr("Searching…")
            Connecting -> L10n.tr("Connecting…")
            Connected -> L10n.tr("Connected")
            is Disconnected -> reason ?: L10n.tr("Disconnected")
        }
}

/// Why a stream start was requested. The cause stays attached if the write waits behind a
/// tare; a start without its reason makes the next hardware trace uninterpretable.
enum class StreamStartCause(val rawValue: String) {
    initial("initial"),
    reconnect("reconnect"),
    foreground("foreground"),
    watchdog("watchdog"),
    tareRecovery("tareRecovery"),
    manualMeasurement("manualMeasurement"),

    /// The human tapped Wake. Distinct from `watchdog`: crediting a manual recovery to the
    /// automatic one would lie in the one log that explains who revived the stream.
    manualWake("manualWake");

    val label: String
        get() = when (this) {
            initial -> "initial"
            reconnect -> "reconnect"
            foreground -> "foreground"
            watchdog -> "watchdog"
            tareRecovery -> "tare recovery"
            manualMeasurement -> "manual measurement"
            manualWake -> "manual wake"
        }
}

/// Why a stream stopped — lets the breadcrumb ring tell a genuine stall from the app doing
/// what it should. **No default anywhere it is taken**: a defaulted cause is wrong at every
/// call site nobody revisited, and mislabelling an automatic stop "stopped by hand" sends
/// the reader looking for a user who was never there.
enum class StreamStopCause(val rawValue: String) {
    sessionEnded("sessionEnded"),
    userStopped("userStopped"),
    measurementComplete("measurementComplete"),
    screenClosed("screenClosed"),
    timedOut("timedOut"),
    disconnecting("disconnecting"),
    sleeping("sleeping");

    val label: String
        get() = when (this) {
            sessionEnded -> "session ended"
            userStopped -> "stopped by hand"
            measurementComplete -> "measurement finished"
            screenClosed -> "screen closed"
            timedOut -> "timed out"
            disconnecting -> "disconnecting"
            sleeping -> "gauge sleeping"
        }
}

/// Client-side lifecycle facts that are not connection-state changes, so the breadcrumb
/// ring can show quarantine boundaries without Bluetooth objects leaving the client.
sealed interface ProgressorClientDiagnostic {
    /// A scanner lifecycle fact, never a device write or an advertisement payload.
    data class BroadcastScan(val event: String) : ProgressorClientDiagnostic
    data object RetiringPeripheral : ProgressorClientDiagnostic
    data object QuarantineReleased : ProgressorClientDiagnostic
    data class StreamStartDeferred(val cause: StreamStartCause) : ProgressorClientDiagnostic
    data class StreamStartWritten(val cause: StreamStartCause) : ProgressorClientDiagnostic

    /// A remotely calibrated gauge's progress to producing force; only from a kind that
    /// `requiresRemoteCalibration`.
    data class Calibration(val status: GaugeCalibrationStatus) : ProgressorClientDiagnostic
}

/// The seam between the app and the gauge.
///
/// **Isolation.** On iOS this protocol is `@MainActor`: CoreBluetooth runs on
/// `queue: .main`, so callbacks arrive on the main thread with no boundary to cross. At ~10
/// notifications/sec of cheap TLV decoding a second isolation domain gains nothing.
///
/// TRANSLATION NOTE: Kotlin has no isolation to propagate, so it is a CONTRACT: Android BLE
/// callbacks arrive on a binder thread, and each client re-posts to
/// `Dispatchers.Main.immediate` before touching state or calling back. `immediate` so a
/// `connect()` that answers synchronously (the mock, a reattach) publishes in the same
/// turn.
///
/// Callbacks rather than Flow, as on iOS: every hop stays visible in one place, and a flow
/// per fact would fan ~80 samples a second out through every collector.
interface ProgressorClient {
    var onEvent: ((ProgressorEvent) -> Unit)?
    var onPacketBoundary: ((PacketBoundary) -> Unit)?
        get() = null
        set(_) {}
    var onStateChange: ((ProgressorConnectionState) -> Unit)?
    var onDiagnostic: ((ProgressorClientDiagnostic) -> Unit)?

    val state: ProgressorConnectionState

    /// Advertised name of the connected unit, e.g. `Progressor_1234`.
    val deviceName: String?

    /// Which device family this client drives. Capability gating (tare, background
    /// streaming, timing source) reads `kind.capabilities`, never the concrete client type.
    ///
    /// TRANSLATION NOTE: Swift's protocol-extension default becomes an interface default,
    /// covering the two Tindeq clients (live and mock); multi-device clients override it.
    val kind: GaugeKind get() = GaugeKind.progressor

    /// Scan for, and connect to, the first gauge found. Idempotent.
    fun connect()

    fun disconnect()

    /// **`startWeightMeasurement` is REFUSED here, on every client**: starts carry a cause
    /// through `startStreaming(cause:)`, so a hardware breadcrumb is never a guess.
    fun send(command: ProgressorCommand)

    fun startStreaming(cause: StreamStartCause)

    fun stopStreaming() {
        send(ProgressorCommand.stopWeightMeasurement)
    }

    fun tare() {
        send(ProgressorCommand.tare)
    }

    fun readBattery() {
        send(ProgressorCommand.getBatteryVoltage)
    }

    /// Sleep the gauge and end the connection without letting an automatic reconnect race
    /// the power-down. Explicit power-down only; normal completion keeps the gauge
    /// available for the second daily session.
    fun sleepDevice() {
        send(ProgressorCommand.enterSleep)
    }
}

/** Notification boundaries are presentation/diagnostic metadata, never engine timing. */
sealed interface PacketBoundary {
    data class Began(val receivedAt: Double) : PacketBoundary
    data object Ended : PacketBoundary
}

internal inline fun ProgressorClient.withPacket(receivedAt: Double, action: () -> Unit) {
    onPacketBoundary?.invoke(PacketBoundary.Began(receivedAt))
    try { action() } finally { onPacketBoundary?.invoke(PacketBoundary.Ended) }
}
